### ISSUE TITLE
Building edited source code locally fails

### ISSUE THREAD

**Issue by samasri** [samasri/openj9#3]
In order to test my changes I did (when virtualizing CodeGenerator), I am building omr locally however it is failing. 
I am trying to build inside docker container built from the [Dockerfile](https://github.com/eclipse/openj9/blob/master/buildenv/docker/jdk9/x86_64/ubuntu16/Dockerfile) indicated by [Nazim's guide](https://github.com/ualberta-smr/nazim/wiki/Week-28:-27th-November-to-1st-December-2017#steps-to-successfully-run-the-linter-for-openj9). Below are more details about my attempts:

# Built openj9 from github/eclipse sources (failed)
* I tried building OpenJ9 by replacing the default sources with _https://github.com/eclipse/omr_ and _https://github.com/eclipse/openj9_ (failed)
* The code I executed is below:
```
git clone https://github.com/ibmruntimes/openj9-openjdk-jdk9.git
cd openj9-openjdk-jdk9/
bash ./get_source.sh
# Change sources
rm -fr omr
git clone https://github.com/eclipse/omr omr
rm -fr openj9
git clone https://github.com/eclipse/openj9 openj9
# Build
bash ./configure --with-freemarker-jar=/root/freemarker.jar
make images | tee out
```
# Build openj9 from default sources and then replace sources (failed)
* I built OpenJ9 with the default sources that come in ./get_source.sh (successfully)
* Then, I replaced them with the _https://github.com/eclipse/omr_ and _https://github.com/eclipse/openj9_
* Tried to build again (failed)
* The code I executed is below:
```
git clone https://github.com/ibmruntimes/openj9-openjdk-jdk9.git
cd openj9-openjdk-jdk9/
# Build with default sources
bash ./get_source.sh
bash ./configure --with-freemarker-jar=/root/freemarker.jar
make images | tee out
# Change sources
rm -fr omr
git clone https://github.com/eclipse/omr omr
rm -fr openj9
git clone https://github.com/eclipse/openj9 openj9
# Rebuild
make images | tee out
```

**Comment by samasri** [comment]
It is normal for the most recent commits from github.com/eclipse/ to temporarily break the openj9 docker build and hence in order to test if my changes are successful I will have to do the following:
- [x] Pick 2 commits: one in github.com/eclipse/omr and the other one in github.com/eclipse/openj9 such that when importing the source code (of omr and openj9) from these 2 commits and building openj9 in docker it builds successfully
- [x] Rebase my OMR and OpenJ9 forks such that my changes are applied directly after the picked commits
- [ ] Verify that when my fork branches are used to build the OpenJ9 project in docker, the project builds successfully [Not completed due to errors]

**Comment by samasri** [comment]
# Step 1:
I picked the following 2 commits: [omr commit](https://github.com/eclipse/omr/commit/42dcf06c22eb292aed5eb72ba7d0058498e8b308) and [openj9 commit](https://github.com/eclipse/openj9/commit/bea496f20be72b37b4c4f26f08c54f803a2207d3)

# Step 2:
I created a new branch called `Virtualizing-CodeGenerator-Testing` on my fork and rebased, I executed the following code:
```
git clone https://github.com/samasri/omr.git omr-test && cd omr-test
git remote add upstream https://github.com/eclipse/omr.git
git fetch upstream
git branch Virtualizing-CodeGenerator-Testing 42dcf06c22eb292aed5eb72ba7d0058498e8b308
git checkout Virtualizing-CodeGenerator
git rebase -i 42dcf06c22eb292aed5eb72ba7d0058498e8b308
git checkout Virtualizing-CodeGenerator-Testing
git rebase -i 42dcf06c22eb292aed5eb72ba7d0058498e8b308
git push origin Virtualizing-CodeGenerator-Testing
```
## Explanation of code
* Line 4: create a branch from the hash of the picked commit
* Line 5: Go to `Virtualizing-CodeGenerator` branch
* Line 6: Locally, rebase that branch to apply the changes on this branch over the picked commit. This step is necessary assuming the `Virtualizing-CodeGenerator` has commits from the upstream repo that happened after the picked commit. For example, assuming that (x*) are commits that happened in the upstream repo and (p*) are commits that are unique for the branch we're on:

`Virtualizing-CodeGenerator` can be: `x1` -> `x2` -> `42dcf06c` -> `x3` -> `x4` -> `p1` -> `p2` -> `p3`
`Virtualizing-CodeGenerator-Test` would be: `x1` -> `x2` -> `42dcf06c`

In that case, we have to apply `p1`, `p2`, and `p3` directly after `42dcf06c`. After rebasing, we will have:

`Virtualizing-CodeGenerator` can be: `x1` -> `x2` -> `42dcf06c` -> `p1` -> `p2` -> `p3`
`Virtualizing-CodeGenerator-Test` would be: `x1` -> `x2` -> `42dcf06c`

* Line 7 and 8: Now we need to rebase `Virtualizing-CodeGenerator-Test` to have `p1`, `p2`, and `p3`.
* Line 9: Lastly, we push the changes to `Virtualizing-CodeGenerator-Test` (but not to `Virtualizing-CodeGenerator`) (we can use `-f` to override any conflicts in the commit history of the branch)

# Step 3
When pushing my new OMR code to the OMR repo PR, it passes successfully, however the issues arise when I try to build the OpenJ9 JDK with my OMR and OpenJ9 source code; it crashes and gives a segmentation fault. In order to understand the problem more, I created 2 new (temporary branches) in each of my OMR and OpenJ9 forks:
* `Test-Branch1`: contains all the commits from IBM developers till the respective commit I picked in step 1.
* `Virtualizing-CodeGenerator-Testing`: contains my changes applied directly on top of `Test-Branch1`.

I also tried to build JDK8 and JDK9, trying to find the source of the problem. Below are the results:

| JDK | OMR Branch | OpenJ9 Branch | Result | Build Log |
| --- | --- | --- | --- | --- |
| 9 | Test-Branch1 | Test-Branch1 | Compiled Successfully | [out](https://github.com/samasri/omr/blob/master/tools/compiler/OMRStatistics/doc/resources/Out%20Logs/jdk9-TestBranch1-TestBranch1) |
| 9 | Virtualizing-CodeGenerator-Testing | Test-Branch1 | Segmentation Fault (when running `make images`) | [out](https://github.com/samasri/omr/blob/master/tools/compiler/OMRStatistics/doc/resources/Out%20Logs/jdk9-VirtualizingCodeGeneratorTesting-TestBranch1) |
| 9 | Virtualizing-CodeGenerator-Testing | Virtualizing-CodeGenerator-Testing | Segmentation Fault (when running `make images`) | [out](https://github.com/samasri/omr/blob/master/tools/compiler/OMRStatistics/doc/resources/Out%20Logs/jdk9-VirtualizingCodeGeneratorTesting-VirtualizingCodeGeneratorTesting) |
| 8 | Test-Branch1 | Test-Branch1 | Compiled Successfully - Functionality tests failed | [out](https://github.com/samasri/omr/blob/master/tools/compiler/OMRStatistics/doc/resources/Out%20Logs/jdk8-TestBranch1-TestBranch1) |
| 8 | Virtualizing-CodeGenerator-Testing | Test-Branch1 | Compiled Successfully - Functionality tests failed | [out](https://github.com/samasri/omr/blob/master/tools/compiler/OMRStatistics/doc/resources/Out%20Logs/jdk8-VirtualizingCodeGeneratorTesting-TestBranch1) |
| 8 | Virtualizing-CodeGenerator-Testing | Virtualizing-CodeGenerator-Testing | Compiled Successfully - Functionality tests failed | [out](https://github.com/samasri/omr/blob/master/tools/compiler/OMRStatistics/doc/resources/Out%20Logs/jdk8-VirtualizingCodeGeneratorTesting-VirtualizingCodeGeneratorTesting) |

**Comment by samasri** [comment]
In order to debug the problems I am getting in step 3, I tried 2 plans:

# Plan 1
Debug the output executable (_build/linux-x86\_64-normal-server-release/jdk/bin/java_):
- Worked with Rob on debugging the Java executable on JDK9. This didn't end up in concrete results; process is described in the [meeting minutes](https://github.com/samasri/omr/wiki/13.-Meeting-Minutes-(19.07.2018)) of our bi-weekly meeting.
- Built JDK8 with debugging symbols (based on this [guide](https://blog.openj9.org/2018/06/05/debugging-openj9-in-docker-with-gdb/)), exactly as instructed (without even changing the omr and openj9 sources). But gdb still reported `no debugging symbols found`; even after copying the debugging symbols to the right directory.
- Tried running the Java executable with different arguments (based on the [problem determination guide](https://github.com/nbhuiyan/omr/blob/8f990ba1ee291063644346e48378d5d3854d2442/doc/compiler/ProblemDetermination.md)) in JDKs 8 and 9.

## Results

PS: 
* In the JDK, there's multiple `java` binary outputs (examples: _jdk/bin/java_, or _images/j2sdk-image/bin/java_). I will be using _images/j2re-image/bin/java_ (based on [Nazim](https://github.com/nbhuiyan/)'s recommendation).
* I always replace the OMR and OpenJ9 source codes with the ones from my fork branches (`Virtualizing-CodeGenerator-Testing`) before building.
* I use WP to signify that I am adding the following extra privileges when running the docker image: `--privileged --cap-add=SYS_PTRACE --security-opt seccomp=unconfined`
* In the java arguments columns (`Xjit`, `Xint`, `Xjit:optLevel=noOpt`, `Xnosigcatch`, `Xint`, `Xnosigcatch`, and `Xrs`), I use 1 to signify success of running java with that version and 0 for fail; the log is found under the Result column.

| JDK | Docker Config | -Xjit | -Xint | -Xjit:optLevel =noOpt | -Xjit: count=0 ,optlevel =hot | -Xnosigcatch -Xint | -Xnosigcatch -Xrs | Result |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
|8|none|0|1|0|0|1|0|[output](https://github.com/samasri/omr/blob/master/tools/compiler/OMRStatistics/doc/resources/Out%20Logs/attemptRunningJDK8Java-virtualized1)
|8|WP|0|1|0|0|1|0|[output](https://github.com/samasri/omr/blob/master/tools/compiler/OMRStatistics/doc/resources/Out%20Logs/attemptRunningJDK8Java-virtualized2)
|9|none|0|1|0|0|1|0|[output](https://github.com/samasri/omr/blob/master/tools/compiler/OMRStatistics/doc/resources/Out%20Logs/attemptRunningJDK8Java-virtualized3)
|9|WP|0|1|0|0|1|0|[output](https://github.com/samasri/omr/blob/master/tools/compiler/OMRStatistics/doc/resources/Out%20Logs/attemptRunningJDK8Java-virtualized4)

# Plan 2

Introduce the changes in multiple phases and see where will the build break. 

## Results

I virtualized all the functions; still have 2 functions that are giving errors:

* `createLinkageForCompilation()`
* `createLinkage(enum TR_LinkageConventions)`

The reason behind the error is document in [samasri/omr:58]().

## Procedure
Built JDK8 with each of my branches in omr and openj9. To build the jdk with a specific branch I did the following:
1. Copied the openj9-openjdk8 to a new directory
2. Ran _.get\_source.sh_
3. Replaced the OMR and OpenJ9 directories with my forks, checking out the right test branch from both OMR and OpenJ9
4. Built the JDK using `make images`
5. Tested `build/linux.../images/j2re-image/bin/java -version`

Created the following branches, where each branch has changes incremented over the branch before it:
* `Test-Branch2`: virtualized one function: `hasDataSnippets`
* `Test-Branch3`: virtualized 10 functions; function names:
    * generateDebugCounterBump
    * branchesAreExpensive
    * canBeAffectedByStoreTagStalls
    * needsNormalizationBeforeShifts
    * codegenSupportsLoadlessBNDCheck
    * isTargetSnippetOrOutOfLine
    * removeUnavailableRegisters
    * emitTargetAddressSnippets
    * nodeMayCauseException
* `Test-Branch4`: virtualized 20 functions; function names:
    * supports32bitAiadd
    * canFoldLargeOffsetInAddressing
    * supportsInliningOfIsInstance
    * opCodeIsNoOpOnThisPlatform
    * shouldValueBeInACommonedNode
    * getSupportsTenuredObjectAlignment
    * isObjectOfSizeWorthAligning
    * getInternalPtrMapBit
    * isPreservedRegister
    * prepareForGRA
    * allowGlobalRegisterAcrossBranch
    * getMaximumNumberOfFPRsAllowedAcrossEdge
    * getMaximumNumbersOfAssignableVRs
    * isRotateAndMask
    * arrayTranslateTableRequiresAlignment
    * arithmeticNeedsLiteralFromPool
    * constLoadNeedsLiteralFromPool
    * supportsHighWordFacility
    * arrayInitMinimumNumberOfBytes
    * isAddressScaleIndexSupported
* `Test-Branch5`: virtualized 40 functions; function names:
    * setNextAvailableBlockIndex
    * getNextAvailableBlockIndex
    * isInternalControlFlowReg
    * startInternalControlFlow
    * endInternalControlFlow
    * supportsDebugCounters
    * getPreferredLoopUnrollFactor
    * getSizeOfCombinedBuffer
    * getMinByteForLongCompareNarrower
    * mulDecompositionCostIsJustified
    * codegenMulDecomposition
    * getSupportsNewObjectAlignment
    * getRegisterMapInfoBitsMask
    * isBranchInstruction
    * isAlignmentInstruction
    * isLabelInstruction
    * splitBlockEntry
    * computeRegisterSaveDescription
    * updateSnippetMapWithRSD
    * setUnavailableRegistersUsage
    * getMaximumNumberOfGPRsAllowedAcrossEdge
    * getMaximumNumbersOfAssignableGPRs
    * considerTypeForGRA
    * excludeInvariantsFromGRAEnabled
    * estimateBinaryLength
    * dumpTargetAddressSnippets
    * apply12BitLabelRelativeRelocation
    * apply16BitLabelRelativeRelocation
    * apply16BitLabelRelativeRelocation
    * apply24BitLabelRelativeRelocation
    * apply32BitLabelRelativeRelocation
    * apply32BitLabelTableRelocation
    * hasTargetAddressSnippets
    * prepareRegistersForAssignment
    * generateNop
    * IsInMemoryType
    * arrayTranslateMinimumNumberOfElements
    * setOnDemandLiteralPoolRun
    * supportsLengthMinusOneForMemoryOpts
    * getSupportsOpCodeForAutoSIMD
* `Test-Branch5.2`: virtualized 10 functions; function names:
    * getGlobalFPRsPreservedAcrossCalls()
    * isReturnInstruction(TR::Instruction *)
    * canNullChkBeImplicit(TR::Node *)
    * isLiteralPoolOnDemandOn()
    * supportsDirectIntegralLoadStoresFromLiteralPool()
    * splitEdge(TR::Instruction *,_Bool,_Bool,TR::Instruction *,TR::list<TR::Instruction *> *,_Bool)
    * supportsDirectJNICallsForAOT()
    * canTransformUnsafeCopyToArrayCopy()
    * CodeGenerator()
    * supportsSinglePrecisionSQRT()
* `Test-Branch5.4`: virtualized 10 functions; function names:
    * usesImplicit64BitGPRs(TR::Node *)
    * setEstimatedLocationsForDataSnippetLabels(int32_t)
    * setCurrentBlockIndex(int32_t)
    * supportsPassThroughCopyToNewVirtualRegister()
    * arrayTranslateAndTestMinimumNumberOfIterations()
    * bitwiseOpNeedsLiteralFromPool(TR::Node *,TR::Node *)
    * ilOpCodeIsSupported(TR::ILOpCodes)
    * materializesLargeConstants()
    * internalPointerSupportImplemented()
    * getMaximumNumberOfGPRsAllowedAcrossEdge(TR::Node *)
* `Test-Branch5.42`: virtualized 1 function: `doRegisterAssignment(enum TR_RegisterKinds)`. **Build crashed here, the reason was that this function has no implementation in the `OMR::CodeGenerator` class. So I declared it as pure virtual.**
* `Test-Branch5.44`: virtualized 1 function: `getGlobalHPRFromGPR(TR_GlobalRegisterNumber)`
* `Test-Branch5.46`: virtualized 1 function: `getCurrentBlockIndex()`
* `Test-Branch5.48`: virtualized 1 function: `GlobalRegisterAvailable(TR_GlobalRegisterNumber,TR::DataType)`
* `Test-Branch5.5`: virtualized 1 function: `supportsTrapsInTMRegion()`
* `Test-Branch5.6`: virtualized 5 functions; function signatures:
    * supportsFusedMultiplyAdd()
    * lowerTreesWalk(TR::Node *,TR::TreeTop *,vcount_t)
    * bndsChkNeedsLiteralFromPool(TR::Node *)
    * getMaximumNumberOfVRFsAllowedAcrossEdge(TR::Node *)
    * enableLiteralPoolRegisterForGRA()
* `Test-Branch5.8`: virtualized 10 functions; function signatures:
    * setEstimatedLocationsForTargetAddressSnippetLabels(int32_t)
    * doBinaryEncoding()
    * doPeephole()
    * getGlobalGPRFromHPR(TR_GlobalRegisterNumber)
    * supportsComplexAddressing()
    * getGlobalRegisterNumber(uint32_t)
    * getSupportsEncodeUtf16BigWithSurrogateTest()
    * buildRegisterMapForInstruction(class TR_GCStackMap *)
    * supportsLongRegAllocation()
    * getSupportsEncodeUtf16LittleWithSurrogateTest()
* `Test-Branch6`: virtualized 10 functions; function signatures:
    * getGlobalGPRsPreservedAcrossCalls()
    * apply64BitLoadLabelRelativeRelocation(TR::Instruction *,TR::LabelSymbol *)
    * isSnippetMatched(TR::Snippet *,int32_t,TR::SymbolReference *)
    * apply32BitLoadLabelRelativeRelocation(TR::Instruction *,TR::LabelSymbol *,TR::LabelSymbol *,int32_t)
    * supportsAtomicAdd()
    * dumpDataSnippets(TR::FILE *)
    * supportsOnDemandLiteralPool()
    * codegenSupportsUnsignedIntegerDivide()
    * isOutOfLineHotPath()
    * beginInstructionSelection
* `Test-Branch6.25`: virtualized 11 functions; function signatures:
    * addProjectSpecializedRelocation(uint8_t *,uint8_t *,uint8_t *,TR_ExternalRelocationTargetKind,char *,uintptr_t,TR::Node *)
    * addProjectSpecializedRelocation(TR::Instruction *,uint8_t *,uint8_t *,TR_ExternalRelocationTargetKind,char *,uintptr_t,TR::Node *)
    * isFenceInstruction(TR::Instruction *)
    * canTransformUnsafeSetMemory()
    * inlineDirectCall(TR::Node *,TR::Register *&)
    * createHWPRecords()
    * incRefCountForOpaquePseudoRegister(TR::Node *,TR::CodeGenerator *,TR::Compilation *)
    * setUpForInstructionSelection()
    * propagateSignThroughBCDLeftShift(TR::DataType)
    * addProjectSpecializedPairRelocation(uint8_t *,uint8_t *,uint8_t *,TR_ExternalRelocationTargetKind,char *,uintptr_t,TR::Node *)
    * lowerTree(TR::Node *,TR::TreeTop *)
* `Test-Branch6.5`: virtualized 9 functions; function signatures:
    * lowerTreesPreChildrenVisit(TR::Node *,TR::TreeTop *,vcount_t)
    * lowerTreesPropagateBlockToNode(TR::Node *)
    * jitAddPicToPatchOnClassUnload(void *,void *)
    * supportsUnneededLabelRemoval()
    * allowSplitWarmAndColdBlocks()
    * generateSwitchToInterpreterPrePrologue(TR::Instruction *,uint8_t,uint8_t)
    * createStackAtlas()
    * needRelocationsForStatics()
    * enableAESInHardwareTransformations()
* `Test-Branch6.6`: virtualized 5 functions; function signatures:
    * jitAdd32BitPicToPatchOnClassUnload(void *,void *)
    * getMinShortForLongCompareNarrower()
    * getMaximumNumbersOfAssignableFPRs()
    * pickRegister(class TR_RegisterCandidate *,TR::Block **,class TR_BitVector &,TR_GlobalRegisterNumber &,TR_LinkHead<class TR_RegisterCandidate> *)
* `Test-Branch6.7`: virtualized 9 functions; function signatures:
    * jitAddUnresolvedAddressMaterializationToPatchOnClassRedefinition(void *)
    * shouldYankCompressedRefs()
    * generateBinaryEncodingPrologue(struct TR_PPCBinaryEncodingData *)
    * suppressInliningOfRecognizedMethod(TR::RecognizedMethod)
* `Test-Branch6.8`: virtualized 4 functions; function signatures:
    * generateCatchBlockBBStartPrologue(TR::Node *,TR::Instruction *)
    * lowerTreeIfNeeded(TR::Node *,int32_t,TR::Node *,TR::TreeTop *)
    * mustGenerateSwitchToInterpreterPrePrologue()
    * allowGuardMerging()
* `Test-Branch6.825`: virtualized one function: `createLinkage(enum TR_LinkageConventions)`. **Running Java seg faulted here because function is being called from a constructor; virtual functions cannot be called from constructors. This function is skipped in the rest of the virtualization.**
* `Test-Branch6.9`: virtualized 4 functions; function signatures:
    * nopsAlsoProcessedByRelocations()
    * alwaysGeneratedSign(TR::Node *)
    * hasComplexAddressingMode()
    * allocateRecompilationInfo()
* `Test-Branch7`: virtualized 5 functions; function signatures:
    * materializesHeapBase()
    * doRematerialization()
    * supportsNegativeFusedMultiplyAdd()
    * jitAddPicToPatchOnClassRedefinition(void *,void *,_Bool)
    * processRelocations()
* `Test-Branch7.1`: virtualized 5 functions; function signatures:
    * yankIndexScalingOp()
    * supportsMergingGuards()
    * processIncomingParameterUsage(class TR_BitVector **,int32_t)
    * lowerTreesPreTreeTopVisit(TR::TreeTop *,vcount_t)
    * getLinkageGlobalRegisterNumber(int8_t,TR::DataType)
* `Test-Branch7.12`: virtualized one function: `createLinkageForCompilation()` **Running Java seg faulted here, same reasoning as `Test-Branch7.825`.**
* `Test-Branch7.2`: virtualized 4 functions; function signatures:
    * getSupportsConstantOffsetInAddressing(int64_t)
    * buildInterpreterEntryPoint()
    * processInstruction(TR::Instruction *,class TR_BitVector **,int32_t &,int32_t &,_Bool)
    * needToAvoidCommoningInGRA()
* `Test-Branch8`: virtualized 20 functions; function signatures:
    * alwaysGeneratesAKnownPositiveCleanSign(TR::Node *)
    * lowerTreesPostChildrenVisit(TR::Node *,TR::TreeTop *,vcount_t)
    * supportsMethodEntryPadding()
    * canUseImmedInstruction(int64_t)
    * hasTMEvaluator()
    * reserveNTrampolines(int32_t)
    * getSupportsBitOpCodes()
    * needClassAndMethodPointerRelocations()
    * wantToPatchClassPointer(const class TR_OpaqueClassBlock *,const uint8_t *)
    * wantToPatchClassPointer(const class TR_OpaqueClassBlock *,const TR::Node *)
    * endInstructionSelection()
    * apply16BitLoadLabelRelativeRelocation(TR::Instruction *,TR::LabelSymbol *,TR::LabelSymbol *,int32_t)
    * jitAdd32BitPicToPatchOnClassRedefinition(void *,void *,_Bool)
    * preLowerTrees()
    * doInstructionSelection()
    * registerAssumptions()
    * alwaysGeneratesAKnownCleanSign(TR::Node *)
    * lowerTreesPostTreeTopVisit(TR::TreeTop *,vcount_t)
    * emitDataSnippets()
    * getMinimumNumberOfNodesBetweenMonitorsForTLE()

The branches are found in OMR and OpenJ9. Most of the cases here are declared in the OMR namespace, and hence the OMR Test-Branch would usually have the added `virtual` keyword in the declaration. However, some of these functions have `self()` calls in the J9 namespace. In that case, the respective Test-Branch would have these `self()` calls removed.

**Comment by samasri** [comment]
Results:
1. I first tried to follow [the guide](https://blog.openj9.org/2018/06/05/debugging-openj9-in-docker-with-gdb/) exactly as instructed (building docker image, getting sources from get_source.sh, and building the jd8) gdb still reported `no debugging symbols found`; even after copying the debugging symbols to the right directory.

**Comment by samasri** [comment]
In summary we ran into the following problems:
* Functions that have a definition in `OMR::X86::CodeGenerator` for example but no declaration in `OMR::CodeGenerator` and they are being used in some of `OMR::CodeGenerator`'s methods through `self()`. When virtualizing the functions and removing `self()` such functions cause errors. The way we solved this is by declaring the function in `OMR::CodeGenerator` as pure virtual.
* Functions that have the same case as above but their implementation is in the `J9` namespace. Declaring such functions pure virtual would cause errors when building OMR independently of J9. We fixed this by adding an empty implementation for the function in OMR.
* After virtualizing the functions, some functions are being called in the constructor of an `CodeGenerator` class. Calling virtual functions in constructors cause no compile error but causes the system to crash when run.
